### PR TITLE
refactor(rewrite): use server sent events

### PR DESCRIPTION
**Pull request by simylein**
This pr switches from web sockets to server sent events for performance reasons. 
It does however not rewrite the translation logic and therefore also not perform partial updates.
I tested it locally and it seemed to work, please let me know if you need anything else.

### CODE REVIEW

File: data/src/f1/f1.service.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+					I: 1,
+				}),
+			);
+			resolve(socket);
+		};
+		socket.onclose = () => {
+			error('web socket git closed');
+			resolve(null);
+		};
+		socket.onerror = (err) => {
```
Comment (Slowlydev) [inline data/src/f1/f1.service.ts:73]: What happens if the Socket setup correctly but crashes afterwards?
- does the backend crash?
- does it start again at some point when it crashed? (new connection, retries)

File: data/src/f1/f1.service.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+			error('web socket encountered an error', err);
+			resolve(null);
+		};
+	});
+};
+
+const negotiation = await negotiate();
```
Comment (Slowlydev) [inline data/src/f1/f1.service.ts:80]: We need a way for negotiation not to fail when doing local development as `data-saver` does not have a negotiate endpoint only a WebSocket and is not a full SingalR implementation.

**Review by Slowlydev — CHANGES_REQUESTED** [review]
Hello 👋🏻
Foremost, thanks for the rewrite to Server Sent Events, I appreciate a lot!

I made some comments on the socket about handling crashes and testing.
Also, I see you added a lot of files, some of which I don't think are necessary, and I would prefer them being stripped to keep the code base small.

I am also not sure if it's necessary to log all the requests, responses and all logs. I mostly worry about performance while flushing is happening and If outbound traffic remains stable during it.

File: .gitignore [hunk 3]
```diff
@@ -6,4 +6,3 @@
 
 # compiled
 target/
-*data.*
```
[the remark this answers is not in the document]
Comment (simylein) [reply .gitignore:9]: Sorry, I couldn't et git to track the `f1-data.sh` file.

File: data/.prettierrc [hunk 4]
```diff
@@ -2,7 +2,7 @@
 	"arrowParens": "always",
 	"semi": true,
 	"printWidth": 120,
-	"singleQuote": false,
```
[the remark this answers is not in the document]
Comment (simylein) [reply data/.prettierrc:5]: Sure

File: data/example.env [hunk 5]
```diff
@@ -1,5 +1,18 @@
 PORT=4000
+NAME=f1-data
 
-# F1_BASE_URL=ws://127.0.0.1:8000 # testing
-F1_BASE_URL=wss://livetiming.formula1.com/signalr
-F1_NEGOTIATE_URL=https://livetiming.formula1.com/signalr
+F1_BASE_URL=livetiming.formula1.com
```
[the remark this answers is not in the document]
Comment (simylein) [reply data/example.env:4]: Haven't thought about that

File: data/package.json [hunk 6]
```diff
@@ -1,15 +1,37 @@
 {
+	"name": "f1-data",
+	"version": "0.4.2",
+	"license": "GPL-3.0-only",
```
[the remark this answers is not in the document]
Comment (simylein) [reply data/package.json:4]: That's just copy pasta pick whatever you like

File: data/src/f1/f1.service.ts (hunk 1 above)
Comment (simylein) [reply data/src/f1/f1.service.ts:73]: Still todo

**Comment by simylein** [comment]
> Hello 👋🏻 Foremost, thanks for the rewrite to Server Sent Events, I appreciate a lot!
> 
> I made some comments on the socket about handling crashes and testing. Also, I see you added a lot of files, some of which I don't think are necessary, and I would prefer them being stripped to keep the code base small.
> 
> I am also not sure if it's necessary to log all the requests, responses and all logs. I mostly worry about performance while flushing is happening and If outbound traffic remains stable during it.

The logging is not necessary, but performance wise it can handle it with ease. 
I just like a robust logging system which we can look back to if things go south

File: data/lib/auth/hash.ts
[the remark this answers is not in the document]
Comment (simylein) [reply data/lib/auth/hash.ts:1]: Will try, but hard while not breaking lib/core

File: data/lib/config/config.ts [hunk 7]
```diff
@@ -1,25 +1,64 @@
...
-			testing: boolean().optional().default(false),
-			logLevel: union(["trace", "debug", "info", "warn", "error"]).optional().default("info"),
+			name: string().optional().max(12).default('f1-data'),
+
+			f1BaseUrl: string().optional().min(4).max(256).default('livetiming.formula1.com'),
+			allowOrigin: string().optional().default('*'),
+			globalPrefix: string().optional().max(12).default(''),
+			defaultVersion: number().optional().transform().min(1).default(0),
+
+			jwtSecret: string().optional().default(randomBytes(32).toString('hex')),
```
[the remark this answers is not in the document]
Comment (simylein) [reply data/lib/config/config.ts]: Will try, but hard while not breaking lib/core

File: data/lib/fake/lorem.ts
[the remark this answers is not in the document]
Comment (simylein) [reply data/lib/fake/lorem.ts:1]: WIll be stripped

File: data/lib/test/expect-type.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export const expectType = <T>(type: T): type is T => true;
```
[the remark this answers is not in the document]
Comment (simylein) [reply data/lib/test/expect-type.ts]: Will be stripped

File: data/src/f1/f1.service.ts (hunk 1 above)
Comment (simylein) [reply data/src/f1/f1.service.ts:73]: Currently does not crash, I can add a reconnect, need to play with it more

File: .gitignore (hunk 3 above)
[the remark this answers is not in the document]
Comment (simylein) [reply .gitignore:9]: Can I do *.data.*?

File: data/lib/core/boot/boot.ts [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+			start(request, 'routing');
+			const matchingRoutes = global.server.routes.filter((route) => compareEndpoint(route, endpoint));
+			const targetRoute = matchingRoutes.find((route) => compareMethod(route, method));
+			stop(request, 'routing');
+
+			const cache = cacheOptions(request, targetRoute);
+			const throttle = throttleOptions(targetRoute);
+
+			if (method === 'options') {
+				const authRoutes = matchingRoutes.filter((route) => route.schema?.jwt);
```
[the remark this answers is not in the document]
Comment (simylein) [reply data/lib/core/boot/boot.ts:55]: It's kinda hard at this point because removing folders is easy but auth is quite nested. I will try

**Review by Slowlydev — APPROVED** [review]
Thanks for all the changes, if logging doesn't affect performance I think we are good and its good to have then.

**Review by Slowlydev — APPROVED** [review]
lgtm 🎉

**Pull request merged**